### PR TITLE
Refactor predictors

### PR DESCRIPTION
**Pull request by danyoungday**
Refactored predictors to be abstracted easily for volunteers to make their own.

Created Predictor abstract class to inherit from with save/load and fit/predict functions.
Created NeuralNetPredictor and abstract SKLearnPredictor which becomes LinearRegressionPredictor and RandomForestPredictor.

Predictors keep track of their own features/scalers/etc. It is on the user to handle this and the evaluator just runs predict on the test set.

Created simple evaluator that loads the test set and evaluates Predictors.

Created unit tests for predictor model types.

Created demo jupyter notebook to show off user flow in order to create and evaluate their predictors.

Note: The changes to data are the same as #60, just slightly refactored. Sorry!

### CODE REVIEW

File: use_cases/eluc/.gitignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+predictors/*/*
```
Comment (danyoungday) [inline use_cases/eluc/.gitignore:1]: This is to prevent pushing the saved models

File: use_cases/eluc/data/eluc_data.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+from data import constants
+
+class ELUCData():
+    """
+    Object to automatically handle the processing of ELUC data.
+    Load with import_data() then process into a df with da_to_df().
+    Maintains train and test dataframes, encoder for data, and encoded versions of train and test.
+    """
+    
+    def __init__(self, path: str, update_path=None, start_year=1851, test_year=2012, end_year=2022, countries=None):
```
Comment (danyoungday) [inline use_cases/eluc/data/eluc_data.py:20]: This allows us to either load with local data or from huggingface

File: use_cases/eluc/data/eluc_data.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        """
+        idx = constants.COUNTRIES_DF[constants.COUNTRIES_DF["abbrevs"].isin(countries)].index.values
+        return df[df["country"].isin(idx)].copy()
+
+    def hf_to_df(self, hf_repo):
+        """
+        Loads dataset from huggingface, converts to pandas, then sets indices appropriately to time/lat/lon.
+        Keep old time/lat/lon columns so we can use them as features later.
+        """
+        ds = load_dataset(hf_repo)["train"]
```
Comment (danyoungday) [inline use_cases/eluc/data/eluc_data.py:55]: As of now, the entire dataset is labeled as "train" on huggingface, we can modify this later.

File: use_cases/eluc/predictors/evaluator.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+from sklearn.metrics import mean_absolute_error
+
+from data.eluc_data import ELUCData
+from predictors.predictor import Predictor
+
+class Evaluator():
```
Comment (danyoungday) [inline use_cases/eluc/predictors/evaluator.py:6]: Very simple evaluator. Holds a test set in memory and then can evaluate a given Predictor object on it. This way, if we end up hiding the test set this is where we can download it.

File: use_cases/eluc/predictors/predictor.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    """
+
+
+    @abstractmethod
+    def fit(self, X_train: pd.DataFrame, y_train: pd.Series):
+        """
+        Fits the model to the training data.
+        :param X_train: dataframe with input data:
+            The input data consists of a dataframe with columns found in the ELUC huggingface repo.
+            It is up to the model to decide which columns to use.
```
Comment (danyoungday) [inline use_cases/eluc/predictors/predictor.py:23]: As discussed earlier.

File: use_cases/eluc/predictors/demo_predictors.ipynb [hunk 6]
```diff
@@ -0,0 +1,197 @@
+{
```
Comment (danyoungday) [inline use_cases/eluc/predictors/demo_predictors.ipynb:1]: This file shows off the workflow. It instantiates a model, fits, predicts, and saves it. Then it instantiates another one, loads, predicts, and compares the 2 outputs to show they are the same.

File: use_cases/eluc/data/eluc_data.py [hunk 7]
```diff
@@ -12,6 +12,7 @@
 from unileaf_util.framework.transformers.data_encoder import DataEncoder
 
 from data import constants
+
```
Comment (danyoungday) [inline use_cases/eluc/data/eluc_data.py:15]: Some random white space changes left over from merging, oops shouldn't matter at all.

File: use_cases/eluc/predictors/evaluator.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+    def evaluate_predictor(self, predictor: Predictor) -> float:
+        """
+        Evaluates a given prescriptor object on a test set.
+        :param predictor: a predictor object.
+        :param X_test: the input data to predict on.
+        :param y_test: the target data to predict.
+        :return: the mean absolute error of the predictor on the test set.
+        """
+        return mean_absolute_error(self.y_test, predictor.predict(self.X_test))
```
Comment (ofrancon) [inline use_cases/eluc/predictors/evaluator.py:26]: Good. We can extend the metrics later, and maybe specify them as a parameter to `evaluate_predictor` or to the constructor.

File: use_cases/eluc/predictors/demo_predictors.ipynb
Comment (ofrancon) [inline use_cases/eluc/predictors/demo_predictors.ipynb:1]: I don't think you've checked in the Predictor implementations yet, but that can come next 👍

**Review by ofrancon — APPROVED** [review]
Looks good to me. I don't see the Predictor implementations yet but that can come later.

File: use_cases/eluc/tests/test_predictors.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+from pathlib import Path
+
+import pandas as pd
+
+from predictors.neural_network.neural_net_predictor import NeuralNetPredictor
+from predictors.sklearn.sklearn_predictor import LinearRegressionPredictor, RandomForestPredictor
+
+class TestPredictors(unittest.TestCase):
+    """
+    Tests the Neural Net predictor.
```
Comment (ofrancon) [inline use_cases/eluc/tests/test_predictors.py:12]: Does more than that - apparently it will test the 3 implementations

File: use_cases/eluc/.gitignore (hunk 1 above)
Comment (danyoungday) [reply use_cases/eluc/.gitignore:1]: Whoops, yes. This ignores all the predictor implementations. I'll use another PR for those but for now there are tests that correspond to models that don't exist yet.

File: use_cases/eluc/predictors/predictor.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        :param y_train: series with target data
+        """
+
+
+    @abstractmethod
+    def predict(self, X_test: pd.DataFrame) -> np.array:
+        """
+        Creates a dataframe with predictions for the input dataframe.
+        :param X_test: dataframe with input data
+        :return: dataframe with predictions
```
[the remark this answers is not in the document]
Comment (danyoungday) [reply use_cases/eluc/predictors/predictor.py:33]: Currently we don't keep track of the name of the label but I can add that in so that we can return the properly named dataframe.
